docs/architecture/adr-047-extend-upgrade-plan.md

```diff
@@ -0,0 +1,188 @@
+
+## Context
+
+The `upgrade` module in conjunction with Cosmovisor are designed to facilitate and automate a blockchain's transition from one version to another.
+
+Users submit a software upgrade governance proposal containing an upgrade `Plan`.
+The `Plan` currently contains the following fields:
+- `name`: A short string identifying the new version.
+- `height`: The chain height at which the upgrade is to be performed.
+- `info`: A string containing information about the upgrade. It can either be a json object (with a specific structure defined through documentation), or it can be a URL that will return such json.
+
+The json object from the `info` field identifies URLs used to download the new blockchain executable for different platforms (OS and Architecture, e.g. "linux/amd64").
+Such a URL can either return the executable file directly or can return an archive containing the executable and possibly other assets.
+If an archive contains a `bin/pre-upgrade` file, Cosmovisor executes it (and waits for it to finish) before attempting to restart the chain.
+
+Currently, there is no mechanism that makes Cosmovisor run a command after the upgraded chain has been restarted.
+
+## Decision
+
+### Protobuf Updates
+
+We will define a new message for providing upgrade instructions and add it as a new field in the `Plan` message.
+The upgrade instructions will contain a list of assets available for each platform.
+It will also allow for the definition of a pre-run and post-run command.
+
+```protobuf
+message Plan {
+  // ... (existing fields)
+
+  UpgradeInstructions upgrade = 6;
+}
+```
+
+The new `UpgradeInstructions upgrade` field MUST be optional.
+
+```protobuf
+message UpgradeInstructions {
+  string pre_run = 1;
+  string post_run = 2;
+  repeated Asset assets    = 3;
+  string description = 4;
+}
+```
+
+All fields in the `UpgradeInstructions` SHOULD be optional.
+- `pre_run` is a command to run prior to the upgraded chain restarting.
+  If defined, this command SHOULD behave the same as the current [pre-upgrade](https://github.com/cosmos/cosmos-sdk/blob/master/docs/migrations/pre-upgrade.md) command.
+  The working directory this command runs from SHOULD be `{DAEMON_HOME}/{upgrade name}`.
+- `post_run` is a command to run after the upgraded chain has been started. If defined, this command SHOULD be only executed once.
+  The output and exit code SHOULD be logged but SHOULD NOT affect the running of the upgraded chain.
+  The working directory this command runs from SHOULD be `{DAEMON_HOME}/{upgrade name}`.
+- `assets` SHOULD allow any number of entries.
+  It SHOULD have only one entry per platform.
+- `description` contains additional information about the upgrade and might contain references to external resources.
+  It SHOULD NOT be used for structured processing information.
+
+```protobuf
+message Asset {
+  string platform = 1;
+  string url = 2;
+  string checksum = 3;
+}
```

Comment:
I'm not sure I like that idea. Again, we'd run into the question of what the working directory is for the commands, and it'd have to be different from the `pre_run` and `post_run` working directories. And since it'd have to be hard-coded, I'm not sure how useful it'd end up being.

Similar behavior could also be achieved just using the `pre_run` command. We'd just need to make sure that, after successfully downloading the asset, the `pre_run` command is run regardless of the existence of `{DAEMON_HOME}/cosmovisor/{upgrade name}/bin/{DAEMON_NAME}`. That way, the asset for the platform can be an archive with a script. It'd be downloaded, then the `pre_run` command would execute that script which could then be responsible for downloading the new binary.